src/app/clusters/commodity-metering-server/commodity-metering-server.cpp

```diff
@@ -0,0 +1,266 @@
+    {
+        for (size_t i = 0; i < a.Value().size(); i++)
+        {
+            if (a.Value()[i] == b.Value()[i])
+            {
+                continue;
+            }
+            return false;
+        }
+        return true;
+    }
+    return false;
+}
+
+template <typename T>
+struct SpanCopier
+{
+    static bool Copy(const chip::Span<const T> & source, DataModel::List<const T> & destination,
+                     Platform::ScopedMemoryBuffer<T> * bufferOut, size_t maxElements = std::numeric_limits<size_t>::max())
+    {
+        if (source.empty())
+        {
+            destination = DataModel::List<const T>();
+            return true;
+        }
+
+        size_t elementsToCopy = std::min(source.size(), maxElements);
+
+        if (!bufferOut->Calloc(elementsToCopy))
+        {
+            return false;
+        }
+
+        std::copy(source.begin(), source.begin() + elementsToCopy, bufferOut->Get());
+        destination = DataModel::List<const T>(chip::Span<const T>(bufferOut->Get(), elementsToCopy));
+        return true;
+    }
+};
+} // namespace
+
+Instance::~Instance()
+{
+    Shutdown();
+}
+
+CHIP_ERROR Instance::Init()
+{
+    mMeteredQuantity.SetNull();
+    mMeteredQuantityTimestamp.SetNull();
+    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INCORRECT_STATE);
+    return CHIP_NO_ERROR;
+}
+
+void Instance::Shutdown()
+{
+    AttributeAccessInterfaceRegistry::Instance().Unregister(this);
+}
+
+CHIP_ERROR Instance::CopyMeteredQuantityEntry(Structs::MeteredQuantityStruct::Type & dest,
+                                              Platform::ScopedMemoryBuffer<uint32_t> * destTariffComponentIDsBuffer,
+                                              const Structs::MeteredQuantityStruct::Type & src)
+{
+    dest.quantity = src.quantity;
+
+    if (!SpanCopier<uint32_t>::Copy(src.tariffComponentIDs, dest.tariffComponentIDs, destTariffComponentIDsBuffer,
+                                    kMaxTariffComponentIDsPerMeteredQuantityEntry))
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::SetMeteredQuantity(const DataModel::Nullable<DataModel::List<Structs::MeteredQuantityStruct::Type>> & newValue)
+{
+    assertChipStackLockedByCurrentThread();
+
+    if (NullableListEqual<Structs::MeteredQuantityStruct::Type>(newValue, mMeteredQuantity))
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    if (mOwnedMeteredQuantityStructBuffer.Get() != nullptr)
+    {
+        mOwnedMeteredQuantityStructBuffer.Free();
+    }
+
+    for (size_t idx = 0; idx < kMaxMeteredQuantityEntries; idx++)
+    {
+        if (mOwnedMeteredQuantityTariffComponentIDsBuffer[idx]->Get() != nullptr)
+        {
+            mOwnedMeteredQuantityTariffComponentIDsBuffer[idx]->Free();
+        }
+    }
+
+    const size_t len = newValue.IsNull() ? 0 : newValue.Value().size();
+
+    if (!newValue.IsNull() && len)
```

Comment:
```suggestion
    if (len)
```

If `newValue.IsNull()`, `len` will be 0.